source/common/config/delta_subscription_state.h

```diff
@@ -22,11 +22,61 @@ namespace Config {
 // There can be multiple DeltaSubscriptionStates active. They will always all be
 // blissfully unaware of each other's existence, even when their messages are
 // being multiplexed together by ADS.
+//
+// There are two scenarios which affect how DeltaSubscriptionState manages the resources. First
+// scenario is when we are subscribed to a wildcard resource, and other scenario is when we are not.
+//
+// Delta subscription state also divides the resources it cached into three categories: requested,
+// wildcard and ambiguous.
+//
+// The "requested" category is for resources that we have explicitly asked for (either through the
+// initial set of resources or through the on-demand mechanism). Resources in this category are in
+// one of two states: "complete" and "waiting for server".
+//
+// "Complete" resources are resources about which the server sent us the information we need (for
+// now - just resource version).
+//
+// The "waiting for server" state is either for resources that we have just requested, but we still
+// didn't receive any version information from the server, or for the "complete" resources that,
+// according to the server, are gone, but we are still interested in them - in such case we strip
+// the information from the resource.
+//
+// The "wildcard" category is for resources that we are not explicitly interested in, but we are
+// indirectly interested through the subscription to the wildcard resource.
+//
+// The "ambiguous" category is for resources that we stopped being interested in, but we may still
+// be interested indirectly through the wildcard subscription - resources in these category are
+// "waiting" for the config server to confirm their status.
+//
+// Please refer to drawings (non-wildcard-resource-state-machine.png and
+// (wildcard-resource-state-machine.png) for visual depictions of the resource state machine.
+//
+// In the "no wildcard subscription" scenario all the cached resources should be in the "requested"
+// category. Resources are added to the category upon the explicit request and dropped when we
+// explicitly unsubscribe from it. Transitions between "complete" and "waiting for server" happen
+// when we receive messages from the server - if a resource in the message is in "added resources"
+// list (thus contains version information), the resource becomes "complete". If the resource in the
+// message is in "removed resources" list, it changes into the "waiting for server" state. If a
+// server sends us a resource that we didn't request, it's going to end up in the "wildcard"
```

Comment:
If the client has not issued a wildcard subscription and the server sends us a resource that we did not request, I think we should actually ignore that resource, because we know that we didn't request it.

From the [How the client specifies what resources to return](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return) section of the xDS spec:

> The client will silently ignore any supplied resources that were not explicitly requested. When the client sends a new request that changes the set of resources being requested, the server must resend any newly requested resources, even if it previously sent those resources without having been asked for them and the resources have not changed since that time.

Reply:
> If the client has not issued a wildcard subscription and the server sends us a resource that we did not request, I think we should actually ignore that resource, because we know that we didn't request it.
> 
> From the [How the client specifies what resources to return](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return) section of the xDS spec:
> 
> > The client will silently ignore any supplied resources that were not explicitly requested. When the client sends a new request that changes the set of resources being requested, the server must resend any newly requested resources, even if it previously sent those resources without having been asked for them and the resources have not changed since that time.

Oh, nice. I can implement that, yes. But I think we haven't ever ignored any resource before, but instead we took into cache everything that came in the message (other than some heartbeat responses).

Reply:
Implemented the ignoring of resources that weren't explicitly asked for when not subscribed to wildcard resource, and added a test.